lfs.c

```diff
@@ -3189,6 +3189,82 @@ static int lfs_rawstat(lfs_t *lfs, const char *path, struct lfs_info *info) {
     return lfs_dir_getinfo(lfs, &cwd, lfs_tag_id(tag), info);
 }
 
+#ifndef LFS_READONLY
+static int lfs_dir_prep_removal(lfs_t *lfs, struct lfs_mlist *dir,
+        lfs_mdir_t *newcwd, uint16_t newid, lfs_block_t *pair,
+        lfs_gstate_t *tmp_gstate)
+{
+    lfs_stag_t res = lfs_dir_get(lfs, newcwd, LFS_MKTAG(0x700, 0x3ff, 0),
+            LFS_MKTAG(LFS_TYPE_STRUCT, newid, 8), pair);
+    if (res < 0) {
+        return (int)res;
+    }
+    lfs_pair_fromle32(pair);
+
+    int err = lfs_dir_fetch(lfs, &dir->m, pair);
+    if (err) {
+        return err;
+    }
+
+    if (dir->m.count > 0 || dir->m.split) {
+        uint16_t id = 0;
+
+        // Walk tags stored in this directory and check for any directory
+        // tags.  Removal of directories with a directory in them can lead
+        // to additional orphans in the filesystem, so we return
+        // LFS_ERR_NOTEMPTY in this case.  Otherwise, leave the loaded
+        // directory for the tail end of the directory split to leave a proper
+        // view of the filesystem after removal.
+        while (true) {
+            if (dir->m.count == id) {
+                if (!dir->m.split) {
+                    // We have iterated through the folder to the last
+                    // tag.
+                    break;
+                }
+
+                // Before we fetch the next block, update our fetched gstate xor
+                lfs_dir_getgstate(lfs, &dir->m, tmp_gstate);
+
+                err = lfs_dir_fetch(lfs, &dir->m, dir->m.tail);
+                if (err) {
+                    return err;
+                }
+
+                id = 0;
+            }
+
+            lfs_stag_t tag = lfs_dir_get(lfs, &dir->m, LFS_MKTAG(0x780, 0x3ff, 0),
+                    LFS_MKTAG(LFS_TYPE_NAME, id, 0), NULL);
+
+            if (tag < 0) {
+                return tag;
+            }
+
+            if (lfs_tag_type3(tag) == LFS_TYPE_DIR) {
+                return LFS_ERR_NOTEMPTY;
+            }
+
+            id += 1;
+        }
+    }
+
+    // mark fs as orphaned
+    err = lfs_fs_preporphans(lfs, +1);
+    if (err) {
+        return err;
+    }
+
+    // I know it's crazy but yes, dir can be changed by our parent's
+    // commit (if predecessor is child)
+    dir->type = 0;
+    dir->id = 0;
+    lfs->mlist = dir;
+
+    return 0;
+}
+#endif
+
 #ifndef LFS_READONLY
 static int lfs_rawremove(lfs_t *lfs, const char *path) {
```

Comment:
POSIX is pretty clear that remove should return ENOTEMPTY if the directory isn't empty. Even though this may seem unhelpful, changing this risks surprising users and an incompatible filesystem.

I think [the idea](https://github.com/littlefs-project/littlefs/issues/43#issuecomment-430950521) of creating a new function for this is a good one. If there's no prior art perhaps `lfs_removeall`?